### PR TITLE
Feat: Give ship

### PR DESCRIPTION
**Pull request by Gods-Righthand**
## Summary
Gives creators the ability to give ships to the player through missions

-----------------------

## Feature Details
This is a partial implementation of  #4399 it allows the giving of ships through missions.

## Usage Examples
It can be used in a mission under `on offer`, `on accept`, etc.. 
```
on <action>
    give ship <ship model> [<name>]
```
If the second name is left out, a civilian ship name will be generated for the gifted ship.
e.g.
```
mission "some mission"
    ...
    on complete
        give ship Sparrow "Mote of Might"
        give ship "Star Barge"
```
indicates that completing the mission will result in the player being given a Sparrow that is named "Mote of Might," and being given a Star Barge with a randomly picked civilian name

## Testing Done
Using the example mission gives two ships, one with a name, and one without a predetermined name, it gets generated later.
#### Example mission
[test.txt](https://github.com/endless-sky/endless-sky/files/5188133/testing.txt)

## Performance Impact
n/a

### CODE REVIEW

**Comment by Zitchas** [comment]
I'm happy to see this appear. Thanks for working on it!

**Comment by Gods-Righthand** [comment]
Yes, I picked it up from Fzzr worked on it some then asked NomadicVolcano for help. He did most of it and didn't walk me through what he did, so with my knowledge of C++ I wasn't able to pick it back up and keep going. Since then it has just been sitting.

**Review by Amazinite — CHANGES_REQUESTED** [review]
Needs to conform to the [style guide](https://endless-sky.github.io/styleguide/styleguide.xml).
* Use tabs, not spaces.
* You do not need to use braces if a loop or conditional body is a single expression. Some of your new code uses extra braces and you've also added them to existing code. (Example:)
```
	if(ship == nullptr) // This can also just be if(!ship)
--	{ // Delete
		return;
--	} // Delete
```
* Variable names should be in camelCase, starting with a lower-case letter.
  * recordDepreciation, not record_depreciation.

**Comment by Gods-Righthand** [comment]
Oh, right, the style guide, thanks for reminding me.

File: source/PlayerInfo.cpp [hunk 1]
```diff
@@ -841,15 +841,32 @@ void PlayerInfo::AddShip(const shared_ptr<Ship> &ship)
...
+// Receive a ship as a gift
+shared_ptr<Ship> PlayerInfo::GiftShip(const Ship *model, const std::string &name, bool record_depriciation)
+{
+	return ReceiveShip(model, name, 0, record_depriciation);
+}
+
+
+
+// Purchase or receive a gift of a ship of the given model and name
+shared_ptr<Ship> PlayerInfo::ReceiveShip(const Ship *model, const string &name, int64_t cost, bool record_depriciation)
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:861]: Why does ReceiveShip return a pointer to the ship? Under what circumstances would this ever return anything aside from the ship that was just given to it? The only instance in which a nullptr would be returned is if the cost was greater than the player had, but gifted ships have no cost. (A nullptr could also be returned if the model doesn't exist, but checking if the model exists is already done by DoGiftShip in MissionAction.)

File: source/MissionAction.h [hunk 2]
```diff
@@ -83,7 +84,8 @@ class MissionAction {
 	Conversation conversation;
 	
 	std::map<const GameEvent *, std::pair<int, int>> events;
-	std::map<const Outfit *, int> gifts;
+	std::vector<std::pair<std::string, std::string>> giftShips;
```
Comment (Amazinite) [inline source/MissionAction.h:87]: giftShips should be a map like giftOutfits. I'd suggest `std::map<const Ship *, std::string>` where the string is the name. Reason being that as this code is written right now, a mission could be written to gift a nonexistent ship (e.g. the mission writer had a typo in the ship name), but this won't throw an error until the mission actually tried to gift the ship, because that is the only point where the string for the ship model is actually converted into the actual ship model.

File: source/PlayerInfo.cpp [hunk 3]
```diff
@@ -841,15 +841,32 @@ void PlayerInfo::AddShip(const shared_ptr<Ship> &ship)
...
 {
 	if(!model)
-		return;
+		return nullptr;
 	
 	int day = date.DaysSinceEpoch();
-	int64_t cost = stockDepreciation.Value(*model, day);
-	if(accounts.Credits() >= cost)
+	if(cost == DEFAULT_COST)
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:867]: This DEFAULT_COST seems clumsy to me. The cost parameter can ONLY ever be one of two things; either it is 0, or it is DEFAULT_COST. We should just be using a boolean. This same boolean can replace recordDepreciation, as the cost is always tied to that boolean right now; if cost is 0, then we know that recordDepreciation is false. If cost is DEFAULT_COST, then we know recordDepreciation is true. So just create a single `gifting` bool that is true for GiftShip and false for BuyShip.

**Review by Amazinite — CHANGES_REQUESTED** [review]

File: source/PlayerInfo.cpp (hunk 3 above)
Comment (Gods-Righthand) [reply source/PlayerInfo.cpp]: Yeah, that makes sense. Made the changes

File: source/MissionAction.cpp [hunk 4]
```diff
@@ -33,7 +33,16 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
...
 
 namespace {
-	void DoGift(PlayerInfo &player, const Outfit *outfit, int count, UI *ui)
+	void DoGiftShip(PlayerInfo &player, const Ship *modelName, const std::string &name, UI *ui)
+	{
+		player.GiftShip(modelName, name);
+		
+		string message = "The " + name + " was added to your fleet!";
+		if(ui)
+			ui->Push(new Dialog(message));
```
Comment (Amazinite) [inline source/MissionAction.cpp:42]: This dialog can only ever be one thing, so I wouldn't bother with storing it in a variable. Just pass it directly to the dialog.
```suggestion
		if(ui)
			ui->Push(new Dialog("The " + name + " was added to your fleet!"));
```

File: source/MissionAction.cpp [hunk 5]
```diff
@@ -186,11 +195,15 @@ void MissionAction::Load(const DataNode &node, const string &missionName)
 			conversation.Load(child);
 		else if(key == "conversation" && hasValue)
 			stockConversation = GameData::Conversations().Get(child.Token(1));
+		else if(key ==  "ship" && hasValue)
+		{
+			giftShips[GameData::Ships().Get(child.Token(1))] = (child.Size() > 2 ? child.Token(2) : "");
+		}
```
Comment (Amazinite) [inline source/MissionAction.cpp:201]: Style: unnecessary braces.
```suggestion
			giftShips[GameData::Ships().Get(child.Token(1))] = (child.Size() > 2 ? child.Token(2) : "");
```

File: source/MissionAction.cpp [hunk 6]
```diff
@@ -202,7 +215,9 @@ void MissionAction::Load(const DataNode &node, const string &missionName)
 		{
 			int count = (child.Size() < 3 ? 1 : static_cast<int>(child.Value(2)));
 			if(count >= 0)
+            {
 				requiredOutfits[GameData::Outfits().Get(child.Token(1))] = count;
+            }
```
Comment (Amazinite) [inline source/MissionAction.cpp:220]: Style: unnecessary braces.
```suggestion
				requiredOutfits[GameData::Outfits().Get(child.Token(1))] = count;
```

File: source/MissionAction.cpp [hunk 7]
```diff
@@ -296,9 +311,14 @@ void MissionAction::Save(DataWriter &out) const
...
-		
-		for(const auto &it : gifts)
+		for(const auto &it : giftShips)
+		{
+			out.Write("ship", it.first->Name(), it.second);
+		}
+		for(const auto &it : giftOutfits)
+		{
 			out.Write("outfit", it.first->Name(), it.second);
+		}
```
Comment (Amazinite) [inline source/MissionAction.cpp:321]: Style: unnecessary braces.
```suggestion
			out.Write("ship", it.first->Name(), it.second);
		for(const auto &it : giftOutfits)
			out.Write("outfit", it.first->Name(), it.second);
```

File: source/MissionAction.cpp [hunk 8]
```diff
@@ -445,14 +465,18 @@ void MissionAction::Do(PlayerInfo &player, UI *ui, const System *destination, co
 		for(const auto &eit : it.second)
 			player.AddSpecialLog(it.first, eit.first, eit.second);
 	
+	for(const auto &it : giftShips)
+	{
+		DoGiftShip(player, it.first, it.second, ui);
+	}
```
Comment (Amazinite) [inline source/MissionAction.cpp:471]: Style: unnecessary braces.
```suggestion
		DoGiftShip(player, it.first, it.second, ui);
```

File: source/MissionAction.cpp [hunk 9]
```diff
@@ -494,7 +518,10 @@ MissionAction MissionAction::Instantiate(map<string, string> &subs, const System
 		int day = it.second.first + Random::Int(it.second.second - it.second.first + 1);
 		result.events[it.first] = make_pair(day, day);
 	}
-	result.gifts = gifts;
+	
+	for(const auto &it : giftShips)
+	result.giftShips[it.first] = !it.second.empty() ? it.second : GameData::Phrases().Get("civilian")->Get();
```
Comment (Amazinite) [inline source/MissionAction.cpp:523]: Format: missing tab.
```suggestion
		result.giftShips[it.first] = !it.second.empty() ? it.second : GameData::Phrases().Get("civilian")->Get();
```

File: source/PlayerInfo.cpp [hunk 10]
```diff
@@ -841,15 +841,32 @@ void PlayerInfo::AddShip(const shared_ptr<Ship> &ship)
...
+// Receive a ship as a gift
+void PlayerInfo::GiftShip(const Ship *model, const std::string &name)
+{
+	return ReceiveShip(model, name, 0, true);
+}
+
+
+
+// Purchase or receive a gift of a ship of the given model and name
+void PlayerInfo::ReceiveShip(const Ship *model, const string &name, int64_t cost, bool gifting)
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:861]: Someone still didn't double check what I posted.
```suggestion
void PlayerInfo::BuyShip(const Ship *model, const std::string &name)
{
	ReceiveShip(model, name, false);
}



// Receive a ship as a gift
void PlayerInfo::GiftShip(const Ship *model, const std::string &name)
{
	ReceiveShip(model, name, true);
}



// Purchase or receive a gift of a ship of the given model and name
void PlayerInfo::ReceiveShip(const Ship *model, const string &name, bool gifting)
```

File: source/PlayerInfo.cpp [hunk 11]
```diff
@@ -841,15 +841,32 @@ void PlayerInfo::AddShip(const shared_ptr<Ship> &ship)
...
 {
 	if(!model)
 		return;
 	
 	int day = date.DaysSinceEpoch();
-	int64_t cost = stockDepreciation.Value(*model, day);
-	if(accounts.Credits() >= cost)
+	if(cost == !gifting)
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:867]: Cost is not a boolean.
```suggestion
	int64_t cost = 0;
	if(!gifting)
```

File: source/PlayerInfo.cpp [hunk 12]
```diff
@@ -841,15 +841,32 @@ void PlayerInfo::AddShip(const shared_ptr<Ship> &ship)
...
+		cost = stockDepreciation.Value(*model, day);
+	if(cost <= 0 || accounts.Credits() >= cost)
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:869]: Unnecessary change.
```suggestion
	if(accounts.Credits() >= cost)
```

File: source/PlayerInfo.cpp [hunk 13]
```diff
@@ -859,13 +876,17 @@ void PlayerInfo::BuyShip(const Ship *model, const string &name)
 		ships.back()->SetIsYours();
 		ships.back()->SetGovernment(GameData::PlayerGovernment());
 		
-		accounts.AddCredits(-cost);
+		if(cost)
+			accounts.AddCredits(-cost);
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:880]: Unnecessary change.
```suggestion
		accounts.AddCredits(-cost);
```

File: source/PlayerInfo.h [hunk 14]
```diff
@@ -28,6 +28,7 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 #include <string>
 #include <utility>
 #include <vector>
+#include <limits>
```
Comment (Amazinite) [inline source/PlayerInfo.h:31]: No longer needed.
```suggestion
```

File: source/PlayerInfo.h [hunk 15]
```diff
@@ -247,6 +249,9 @@ class PlayerInfo {
 	
 	
 private:
+	// Internal implementation of BuyShip and GiftShip
+	void ReceiveShip(const Ship *model, const std::string &name, int64_t cost, bool gifting);
```
Comment (Amazinite) [inline source/PlayerInfo.h:253]:
```suggestion
	void ReceiveShip(const Ship *model, const std::string &name, bool gifting);
```

File: source/PlayerInfo.h [hunk 16]
```diff
@@ -325,6 +330,8 @@ class PlayerInfo {
 	Depreciation stockDepreciation;
 	std::set<std::pair<const System *, const Outfit *>> harvested;
 	
+	bool gifting = true;
+
```
Comment (Amazinite) [inline source/PlayerInfo.h:334]: Unnecessary change.
```suggestion
```

**Review by Amazinite — CHANGES_REQUESTED** [review]

File: source/PlayerInfo.cpp [hunk 17]
```diff
@@ -841,14 +841,32 @@ void PlayerInfo::AddShip(const shared_ptr<Ship> &ship)
...
+void PlayerInfo::ReceiveShip(const Ship *model, const string &name, bool gifting)
 {
 	if(!model)
 		return;
 	
 	int day = date.DaysSinceEpoch();
-	int64_t cost = stockDepreciation.Value(*model, day);
+	int64_t cost = 0;
+	if(!gifting)
+		cost = stockDepreciation.Value(*model, day);
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:869]: Just one more quick suggestion.
```suggestion
	int64_t cost = gifting ? 0 : stockDepreciation.Value(*model, day);
```

**Review by Amazinite — CHANGES_REQUESTED** [review]

**Comment by Amazinite** [comment]
Somehow. a commit on this branch had changed files in the util and test folders, but it was one of those strange instances where somethings is changed but at the same time it says file not changed. Couldn't just replace them with "unchanged" files because it didn't recognize that a change had occurred. I opted to just reset the branch to before the commit was made that changes the util and test folder files then update it from there, but turns out that ate every single commit that was here previously.

Need to credit Fzzr and Nomadic once this is merged.

File: source/MissionAction.cpp [hunk 18]
```diff
@@ -33,7 +33,15 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 using namespace std;
 
 namespace {
-	void DoGift(PlayerInfo &player, const Outfit *outfit, int count, UI *ui)
+	void DoGiftShip(PlayerInfo &player, const Ship *modelName, const std::string &name, UI *ui)
```
Comment (petervdmeer) [inline source/MissionAction.cpp:36]: The const Ship pointer points to a complete model, not just to a modelname, so I suggest to call the variable model.

File: source/MissionAction.cpp [hunk 19]
```diff
@@ -186,11 +194,13 @@ void MissionAction::Load(const DataNode &node, const string &missionName)
 			conversation.Load(child);
 		else if(key == "conversation" && hasValue)
 			stockConversation = GameData::Conversations().Get(child.Token(1));
+		else if(key ==  "ship" && hasValue)
+			giftShips[GameData::Ships().Get(child.Token(1))] = (child.Size() > 2 ? child.Token(2) : "");
```
Comment (Amazinite) [inline source/MissionAction.cpp:198]: `>=` is typically used over just `>` for checking the size of a DataNode.
```suggestion
			giftShips[GameData::Ships().Get(child.Token(1))] = child.Size() >= 3 ? child.Token(2) : "";
```

**Review by Amazinite — CHANGES_REQUESTED** [review]
The new comments in PlayerInfo could be improved, and ReceiveShip should be moved in the .h file to somewhere more sensible than above the constructor and in the .cpp file to match its location in the .h file.

File: source/MissionAction.cpp (hunk 19 above)
Comment (petervdmeer) [inline source/MissionAction.cpp]: Should we print a warning when trying to give a ship that doesn't exist in the GameData?

File: source/MissionAction.cpp [hunk 20]
```diff
@@ -186,11 +194,13 @@ void MissionAction::Load(const DataNode &node, const string &missionName)
...
 		else if(key == "outfit" && hasValue)
 		{
 			int count = (child.Size() < 3 ? 1 : static_cast<int>(child.Value(2)));
 			if(count)
-				gifts[GameData::Outfits().Get(child.Token(1))] = count;
+				giftOutfits[GameData::Outfits().Get(child.Token(1))] = count;
```
Comment (petervdmeer) [inline source/MissionAction.cpp:203]: Should we print a warning when trying to give an outfit that doesn't exist in the GameData?

File: source/PlayerInfo.cpp [hunk 21]
```diff
@@ -841,14 +841,30 @@ void PlayerInfo::AddShip(const shared_ptr<Ship> &ship)
 
 
 
-// Buy a ship of the given model, and give it the given name.
-void PlayerInfo::BuyShip(const Ship *model, const string &name)
+// Charge the player the cost of a ship and give it the specified name:
+void PlayerInfo::BuyShip(const Ship *model, const std::string &name)
```
Comment (petervdmeer) [inline source/PlayerInfo.cpp:845]: Why adding "std::" here (and on line 853)? std should be covered by an using statement at the top of the class.

File: source/PlayerInfo.cpp [hunk 22]
```diff
@@ -841,14 +841,30 @@ void PlayerInfo::AddShip(const shared_ptr<Ship> &ship)
...
+
+// Receive a ship as a gift
+void PlayerInfo::GiftShip(const Ship *model, const std::string &name)
+{
+	ReceiveShip(model, name, true);
+}
+
+
+
+// Purchase or receive a gift of a ship of the given model and name
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:860]:
```suggestion
// Adds a ship of the given model with the given name to the player's fleet. If this ship is being gifted, it costs nothing and starts fully depreciated.
```

File: source/MissionAction.cpp (hunk 19 above)
Comment (Gods-Righthand) [reply source/MissionAction.cpp:198]: Changed it in pursuit of a different problem, didn't know it was used more commonly this way.

File: source/PlayerInfo.cpp [hunk 23]
```diff
@@ -841,14 +841,30 @@ void PlayerInfo::AddShip(const shared_ptr<Ship> &ship)
...
-void PlayerInfo::BuyShip(const Ship *model, const string &name)
+// Charge the player the cost of a ship and give it the specified name:
+void PlayerInfo::BuyShip(const Ship *model, const std::string &name)
+{
+	ReceiveShip(model, name, false);
+}
+
+
+
+// Receive a ship as a gift
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:852]:
```suggestion
// Receive a ship of the give model with the given name as a gift.
```

File: source/PlayerInfo.cpp [hunk 24]
```diff
@@ -841,14 +841,30 @@ void PlayerInfo::AddShip(const shared_ptr<Ship> &ship)
 
 
 
-// Buy a ship of the given model, and give it the given name.
-void PlayerInfo::BuyShip(const Ship *model, const string &name)
+// Charge the player the cost of a ship and give it the specified name:
```
Comment (Amazinite) [inline source/PlayerInfo.cpp:844]: This comment doesn't need to change.
```suggestion
// Buy a ship of the given model, and give it the given name.
```

File: source/PlayerInfo.cpp [hunk 25]
```diff
@@ -841,14 +841,30 @@ void PlayerInfo::AddShip(const shared_ptr<Ship> &ship)
...
+// Purchase or receive a gift of a ship of the given model and name
+void PlayerInfo::ReceiveShip(const Ship *model, const string &name, bool gifting)
 {
 	if(!model)
 		return;
 	
 	int day = date.DaysSinceEpoch();
-	int64_t cost = stockDepreciation.Value(*model, day);
+	int64_t cost = gifting ? 0 : stockDepreciation.Value(*model, day);
 	if(accounts.Credits() >= cost)
```
Comment (petervdmeer) [inline source/PlayerInfo.cpp:868]: I believe giving a ship should still work if the player has negative credits. (Change the if-statement to what is is now, but add an or-statement that makes this true when gifting is true.)

File: source/PlayerInfo.h [hunk 26]
```diff
@@ -247,6 +248,9 @@ class PlayerInfo {
 	
 	
 private:
+	// Internal implementation of BuyShip and GiftShip
+	void ReceiveShip(const Ship *model, const std::string &name, bool gifting);
+
```
Comment (Amazinite) [inline source/PlayerInfo.h:253]: Perhaps move this to above or below SelectShip since both functions relate to ships. I'd also change the comment to match the one in the .cpp file.

File: source/PlayerInfo.cpp [hunk 27]
```diff
@@ -863,9 +879,12 @@ void PlayerInfo::BuyShip(const Ship *model, const string &name)
...
 		// Record the transfer of this ship in the depreciation and stock info.
-		depreciation.Buy(*model, day, &stockDepreciation);
-		for(const auto &it : model->Outfits())
-			stock[it.first] -= it.second;
+		if(!gifting)
+		{
+			depreciation.Buy(*model, day, &stockDepreciation);
+			for(const auto &it : model->Outfits())
+				stock[it.first] -= it.second;
+		}
```
Comment (petervdmeer) [inline source/PlayerInfo.cpp:887]: Are ships that are gifts valued at 100% of the deprecation value, or at 25% of the deprecation value? (Or at some other value.)

**Review by Amazinite — CHANGES_REQUESTED** [review]

File: source/PlayerInfo.h [hunk 28]
```diff
@@ -121,8 +121,9 @@ class PlayerInfo {
 	std::map<const std::shared_ptr<Ship>, std::vector<std::string>> FlightCheck() const;
 	// Add a captured ship to your fleet.
 	void AddShip(const std::shared_ptr<Ship> &ship);
-	// Buy or sell a ship.
+	// Buy a ship, receive a gifted ship, or sell a ship.
```
Comment (petervdmeer) [inline source/PlayerInfo.h:124]: Receive a ship given. (a gifted ship is a bit different than a ship that is given to the player).

File: source/PlayerInfo.cpp [hunk 29]
```diff
@@ -841,14 +841,30 @@ void PlayerInfo::AddShip(const shared_ptr<Ship> &ship)
...
+// Charge the player the cost of a ship and give it the specified name:
+void PlayerInfo::BuyShip(const Ship *model, const std::string &name)
+{
+	ReceiveShip(model, name, false);
+}
+
+
+
+// Receive a ship as a gift
+void PlayerInfo::GiftShip(const Ship *model, const std::string &name)
```
Comment (petervdmeer) [inline source/PlayerInfo.cpp:853]: Do we need the 2 functions with the oneliners inside them? I think it looks nicer if we just keep using buyShip as function with a third parameter (isGift) which by default is false. You can give such default parameter in the header.

File: source/PlayerInfo.cpp (hunk 23 above)
Comment (Gods-Righthand) [reply source/PlayerInfo.cpp:852]:
```suggestion
// Receive a ship of the given model with the given name as a gift.
```
the give model or given model?

File: source/MissionAction.cpp (hunk 19 above)
Comment (Amazinite) [reply source/MissionAction.cpp]: Don't we already get an "(outfit | ship) X is referred to but never defined" warning?

File: source/PlayerInfo.cpp (hunk 27 above)
Comment (Amazinite) [reply source/PlayerInfo.cpp:887]: If a ship or outfit how no recorded depreciation, it is fully depreciated.
https://github.com/endless-sky/endless-sky/blob/8f29a0b3d1401d4fd0f4de6e7433867c9dbe7710/source/Depreciation.cpp#L86-L87

File: source/MissionAction.cpp (hunk 19 above)
Comment (petervdmeer) [reply source/MissionAction.cpp]: Good point! Yes that is probably handled by a generic check on empty ship definitions after loading.

File: source/MissionAction.cpp (hunk 20 above)
Comment (petervdmeer) [reply source/MissionAction.cpp:203]: Same as above, probably already handled by a generic check.

File: source/PlayerInfo.cpp (hunk 21 above)
Comment (Gods-Righthand) [reply source/PlayerInfo.cpp:845]:
```suggestion
void PlayerInfo::BuyShip(const Ship *model, const string &name)
```
Not sure, will remove it.

File: source/MissionAction.cpp (hunk 18 above)
Comment (Gods-Righthand) [reply source/MissionAction.cpp:36]:
```suggestion
	void DoGiftShip(PlayerInfo &player, const Ship *model, const std::string &name, UI *ui)
```
Ok, will change it then.

File: source/PlayerInfo.cpp (hunk 22 above)
Comment (Gods-Righthand) [reply source/PlayerInfo.cpp:860]: That adds a lot of clarity, so yes.

File: source/MissionAction.cpp (hunk 19 above)
Comment (Gods-Righthand) [reply source/MissionAction.cpp]: So I don't need to worry about trying to add a checks here? Nice!

File: source/MissionAction.cpp [hunk 30]
```diff
@@ -33,7 +33,15 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 using namespace std;
 
 namespace {
-	void DoGift(PlayerInfo &player, const Outfit *outfit, int count, UI *ui)
+	void DoGiftShip(PlayerInfo &player, const Ship *model, const std::string &name, UI *ui)
+	{
+		player.GiftShip(modelName, name);
```
Comment (Amazinite) [inline source/MissionAction.cpp:38]: You also need to rename the variable here.
```suggestion
		player.GiftShip(model, name);
```

**Review by Amazinite — CHANGES_REQUESTED** [review]

**Review by Amazinite — APPROVED** [review]

File: source/MissionAction.cpp [hunk 31]
```diff
@@ -186,11 +196,13 @@ void MissionAction::Load(const DataNode &node, const string &missionName)
 			conversation.Load(child);
 		else if(key == "conversation" && hasValue)
 			stockConversation = GameData::Conversations().Get(child.Token(1));
+		else if(key ==  "ship" && hasValue)
```
Comment (tehhowch) [inline source/MissionAction.cpp:199]: `ship "some model"` is not a sufficient interface for this functionality.

Unlike outfits, there is not a simple way to reflect the direction of the transfer (e.g. by passing a negative count).

Please refactor this to account for  all of the desired ship transference mechanisms (e.g. give a ship, take a ship, has a ship).

I'm certain several content creators can work with you to determine a better, more flexible syntax.

**Review by tehhowch — CHANGES_REQUESTED** [review]

File: source/MissionAction.cpp (hunk 31 above)
Comment (Amazinite) [reply source/MissionAction.cpp:199]: `ship <action> "<model>" [<name>]`?

Comment (Gods-Righthand) [reply source/MissionAction.cpp:199]: That sounds like a resonable way to put it.

Comment (tehhowch) [reply source/MissionAction.cpp:199]: A drawback to that syntax is the inability to change the implementation to store instantiated ship objects without breaking compatibility.

I think using the action first allows us more flexibility for future improvements.

Comment (Amazinite) [reply source/MissionAction.cpp:199]: What?

Comment (Zitchas) [reply source/MissionAction.cpp:199]: So, something like this? (Based on Tehhowch's suggestion on discord
```
(give|take|require)
  ship <model name> [<name>] [<InvisibleIdentifier>] [<LockShip>] [<LockOutfits>]
  outfit <outfit name> [#count]
(give|take|require) [....inline specifier]
```
InvisibleIdentifier = identifier that is stored invisibly in the ship's attributes so that the player can change the ship's name and still be identified by the mission
LockShip = variable the specifies that the ship's name is locked. (recorded in the ship's attributes)
LockOutfit = variable that specifies that the ship's outfits cannot be modified (recorded in the ship's attributes)

**Comment by Amazinite** [comment]
The syntax is now `give ship "<model>" ["<name>"]`, leaving it open to expansion to include `outfit`, uniquely defined ships, and more actions.

**Review by tehhowch — COMMENTED** [review]
We could optionally add a comment in `CanBeDone` about parsing the gifted ship collection, but it's likely we will be changing things up anyway for the require/take scenarios anyway, so it isn't necessary to do so.

File: source/MissionAction.cpp [hunk 32]
```diff
@@ -33,6 +33,16 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
...
 
 namespace {
+	void DoGift(PlayerInfo &player, const Ship *model, const string &name, UI *ui)
+	{
+		if(model->ModelName().empty())
+			return;
+		
+		player.BuyShip(model, name, true);
+		if(ui)
+			ui->Push(new Dialog("The " + name + " was added to your fleet!"));
```
[the remark this answers is not in the document]
Comment (Amazinite) [reply source/MissionAction.cpp:43]: Perhaps `“A <model>, the <name>, was added to your fleet.”`?

[the remark this answers is not in the document]
Comment (tehhowch) [reply source/MissionAction.cpp:43]: I think `The <model> <noun> "<name>" was added to your fleet.` or `You received the <model> <noun> "<name>."` would work.

To match outfit messaging, it could be `A/An <noun>, the <model> "<name>," was added to your fleet.`

[the remark this answers is not in the document]
Comment (Gods-Righthand) [reply source/MissionAction.cpp]: Based on community feedback on Discord as well, I went with `A/An <noun>, the <model>"<name>," was added to the fleet.`

File: source/MissionAction.cpp [hunk 33]
```diff
@@ -33,6 +33,23 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
...
+		player.BuyShip(model, name, true);
+		
+		string noun = model->Noun();
+		string message;
+		char c = tolower(noun.front());
+		bool isVowel = (c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u');
+		message = (isVowel ? "An " : "A ");
+		message += noun;
+		if(ui)
+			ui->Push(new Dialog(message + ", the " + model->ModelName() + " \"" + name + ",\" was added to your fleet!"));
```
Comment (Amazinite) [inline source/MissionAction.cpp:50]:
```suggestion
			ui->Push(new Dialog(message + ", the " + model->ModelName() + " \"" + name + ",\" was added to your fleet."));
```

**Review by tehhowch — COMMENTED** [review]
For outfits, a dialog is only displayed for an outfit that could not be installed, and in all cases a message is printed in the "chat" log.

In contrast, ship gifting as proposed here will always trigger a dialog per ship. Is giving the player a ship of such importance that a dialog should be displayed for each ship given, **_in addition to_** whatever existing conversation or dialog text is already associated with the given action? Or should the "give" case be only recorded in the message log, and the mission writer expected to have used normal expository methods to let the player know that they got something?

**Comment by Amazinite** [comment]
I'd say the latter.

**Comment by Zitchas** [comment]
I agree with the later. I'd say using this falls into the same category as the payment mechanism: We *can* give people money without mentioning it in conversation, but we generally should tell them about it.

(I"m just imagining a situation where the player goes into the starport, then leaves and they've got this new ship sitting in the shipyard. "Who are you?" "I'm the pilot of your new ship, Sir!" "Where did you come from?" "I just got hired this morning to fly this new ship over to the landing pad here for the new owner, which according to these records, is you, as of today." Player: ???? )

That could be a funny way to start a mission sequence.

File: source/MissionAction.cpp [hunk 34]
```diff
@@ -33,6 +33,15 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 using namespace std;
 
 namespace {
+	void DoGift(PlayerInfo &player, const Ship *model, const string &name, UI *ui)
+	{
+		if(model->ModelName().empty())
+			return;
+		
+		player.BuyShip(model, name, true);
+		Messages::Add("The " + model->ModelName() + " \"" + name + ",\" was added to your fleet.");
```
Comment (tehhowch) [inline source/MissionAction.cpp:42]: Either a comma both before & after the name, or no comma at all

File: source/MissionAction.cpp [hunk 35]
```diff
@@ -186,11 +195,18 @@ void MissionAction::Load(const DataNode &node, const string &missionName)
 			conversation.Load(child);
 		else if(key == "conversation" && hasValue)
 			stockConversation = GameData::Conversations().Get(child.Token(1));
+		else if(key == "give" && hasValue)
+		{
+			if(child.Token(1) == "ship" && child.Size() >= 3)
+				giftShips[GameData::Ships().Get(child.Token(2))] = child.Size() >= 4 ? child.Token(3) : "";
+			else
+				child.PrintTrace("Skipping unsupported give syntax:");
```
Comment (tehhowch) [inline source/MissionAction.cpp:203]:
```suggestion
				child.PrintTrace("Skipping unsupported \"give\" syntax:");
```

**Review by tehhowch — COMMENTED** [review]
Fix these and it'll be in good shape

File: source/MissionAction.cpp (hunk 34 above)
Comment (Gods-Righthand) [reply source/MissionAction.cpp:42]:
```suggestion
		Messages::Add("The " + model->ModelName() + " \"" + name + "\" was added to your fleet.");
```
Okay, whatever.

**Review by tehhowch — APPROVED** [review]

**Comment by Zitchas** [comment]
Question: Is the example file and code in the initial post accurate as to the final formatting we should be using?

**Comment by Amazinite** [comment]
The usage example in the PR description is accurate.

**Comment by Zitchas** [comment]
````
        give ship Sparrow "Mote of Might"
        give ship "Star Barge"
````

Just to clarify, the two options are "give ship with a specified name" and "give a ship with a random name", correct? Neither option allows the player to pick a name on the spot?

**Comment by Amazinite** [comment]
Correct.

**Comment by Zitchas** [comment]
OK. It would be nice to have that functionality, but that's a luxury, not a necessity. The player can manually change the name later. 

Thanks to everyone for working on this and getting it implemented. 

Just for reference, putting it into use: https://github.com/endless-sky/endless-sky/pull/5388
